Create solution for Grass Planting (#797)
cpinitiative/usaco-guide: solutions/USACO jan19-grass-planting.mdx

```java
import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.StringTokenizer;

public class planting {
    public static void main(String[] args) throws IOException {
        BufferedReader f = new BufferedReader(new FileReader("planting.in"));
        PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("planting.out")));
        StringTokenizer st = new StringTokenizer(f.readLine());
        int n = Integer.parseInt(st.nextToken());
        if (n==1) {
            pw.println(1);
            pw.close();
        } else if (n==2) {
            pw.println(2);
            pw.close();
        }
        int[] adj = new int[n];
        Arrays.fill(adj, 1);
        int max=1;
        for (int i = 0; i < n-1; i++) {
            st = new StringTokenizer(f.readLine());
            int ne = Integer.parseInt(st.nextToken())-1;
            if (++adj[ne]>adj[max]) max=ne;
            ne = Integer.parseInt(st.nextToken())-1;
            if (++adj[ne]>adj[max]) max=ne;
        }
        pw.println(adj[max]);
        pw.close();
    }
}
```

## Solution 2 (DFS, not intended)

> This can also be solved using DFS in $\mathcal{O}(V+E)=\mathcal{O}(N+N-1)=\mathcal{O}(2N-1)$, which is linear time (although slightly slower than the previous solution), enough to pass for all testcases. The implementation is straight forward. My java code for this is below:

Reviewer: isn't it still O(N)

Reviewer 2: What happened to [this commit](https://github.com/cpinitiative/usaco-guide/pull/797/commits/55648f37cf16f33e20cb4b0e52bc00a54e8e7319)?

Author: I thought I accepted it...

Reviewer 2: You did -- it seems to have gotten reverted sometime later though..

        StringTokenizer st = new StringTokenizer(f.readLine());
        int farms = Integer.parseInt(st.nextToken());
        adj = new ArrayList<>();
        for (int i = 0; i < farms; i++) {
            adj.add(new ArrayList<Integer>());
        }
        for (int i = 0; i < farms - 1; i++) {
            st = new StringTokenizer(f.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            adj.get(a - 1).add(b - 1);
            adj.get(b - 1).add(a - 1);
        }
        visited = new int[farms];
        Arrays.fill(visited, 0);
        visited[0] = 1;
        dfs(0, 0);
        int max = 0;
        for (int i : visited) {
            max = Math.max(max, i);
        }
        pw.println(max);
        pw.close();
    }

    public static void dfs(int self, int parent) {
        int len = adj.get(self).size();
        int g = 1;
        for (int i : adj.get(self)) {
            if (i != parent) {
                while (g == visited[self] || g == visited[parent]) g++;
                visited[i] = g;
                dfs(i, self);
                g++;
            }
        }
    }
}

> ```

Reviewer 3: For the rest is ok. Also I think you have made two files, @frodakcin  already mentions this.

id: usaco-894
source: USACO Silver 2019 January
title: Grass Planting
author: Neo Wang, Aadit Ambadkar
---

[Official Analysis](http://www.usaco.org/current/data/sol_planting_silver_jan19.html)

## Solution 1 - Intended Solution

**Time Complexity:** $\mathcal{O}(N)$

> Let $\texttt{adj}[i]$ represent the number of paths adjacent to node $i$ in the graph. Then, considering that node, and that node alone, you would need exactly $\texttt{adj}[i]+1$ numbers to fill it. This is because you would need $\texttt{adj}[i]$ for each of the adjacent nodes, and $1$ for the node itself. So, our answer is the maximum of all the adjacents, plus $1$. 

Reviewer 2: ```suggestion
Let $\texttt{deg}[i]$ represent the degree of node $i$: the number of paths connecting the node. Then, considering that node, and that node alone, you would need no less than $\texttt{deg}[i]+1$ grass types. This is because you would need $\texttt{deg}[i]$ for each of the adjacent nodes, and $1$ for the node itself. It can be shown that the tree can also be colored accordingly in $\max(\texttt{deg}[i])+1$ colors. 
```

> Let $\texttt{deg}[i]$ represent the degree of node $i$: the number of paths connecting the node. Then, considering that node, and that node alone, you would need no less than $\texttt{deg}[i]+1$ grass types. This is because you would need $\texttt{deg}[i]$ for each of the adjacent nodes, and $1$ for the node itself. It can be shown that the tree can also be colored accordingly in $\max(\texttt{deg}[i])+1$ colors. 

Reviewer 4: ```suggestion
Let $\texttt{deg}[i]$ represent the degree of node $i$: the number of paths connecting the node. Then, considering that node, and that node alone, you would need $\texttt{deg}[i]+1$ grass types. This is because you would need $\texttt{deg}[i]$ for each of the adjacent nodes, and $1$ for the node itself. It can be shown that the tree can also be colored accordingly in $\max(\texttt{deg}[i])+1$ colors. 
```

</JavaSection>

</LanguageSection>

## Solution 2 - DFS

**Time Complexity**: $\mathcal{O}(N)$

> This can also be solved using DFS. The solution is less elegant but generates a valid assignment of grass types in the $\texttt{deg}$ array. 

Reviewer 4: ```suggestion
This can also be solved using DFS. The solution is less elegant but generates a valid assignment of grass types in the $\texttt{ans}$ array. 
```

> This can also be solved using DFS. The solution is less elegant, but generates a valid assignment of grass types in the $\texttt{visited}$ array. 

Reviewer 4: ```suggestion
This can also be solved using DFS. The solution is less elegant but generates a valid assignment of grass types in the $\texttt{deg}$ array. 
```

<LanguageSection>

<JavaSection>

```java
import java.io.*;
import java.util.*;

public class Main {

> 	public static int[] deg;
> 	public static List<Integer>[] adj;
> 	public static void main(String[] args) throws IOException {
> 		BufferedReader f = new BufferedReader(new FileReader("planting.in"));
> 		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("planting.out")));
> 		StringTokenizer st = new StringTokenizer(f.readLine());
> 		int farms = Integer.parseInt(st.nextToken());
> 		adj = new List[farms];
> 		for (int i = 0; i < farms; i++) {
> 			adj[i] = new ArrayList<>();
> 		}
> 		for (int i = 0; i < farms - 1; i++) {
> 			st = new StringTokenizer(f.readLine());
> 			int a = Integer.parseInt(st.nextToken());
> 			int b = Integer.parseInt(st.nextToken());
> 			adj[a - 1].add(b - 1);
> 			adj[b - 1].add(a - 1);
> 		}
> 		deg = new int[farms];
> 		deg[0] = 1;
> 		dfs(0, 0);
> 		int max = 0;
> 		for (int i : deg) {
> 			max = Math.max(max, i);
> 		}
> 		pw.println(max);
> 		pw.close();
> 	}
> 
> 	public static void dfs(int self, int parent) {
> 		int len = adj.get(self).size();
> 		int g = 1;
> 		for (int i : adj.get(self)) {
> 			if (i != parent) {
> 				while (g == deg[self] || g == deg[parent]) g++;
> 				deg[i] = g;
> 				dfs(i, self);
> 				g++;
> 			}
> 		}
> 	}
> }

Reviewer 4: ```suggestion
	public static int[] ans;
	public static List<Integer>[] adj;
	public static void main(String[] args) throws IOException {
		BufferedReader f = new BufferedReader(new FileReader("planting.in"));
		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("planting.out")));
		StringTokenizer st = new StringTokenizer(f.readLine());
		int farms = Integer.parseInt(st.nextToken());
		adj = new List[farms];
		for (int i = 0; i < farms; i++) {
			adj[i] = new ArrayList<>();
		}
		for (int i = 0; i < farms - 1; i++) {
			st = new StringTokenizer(f.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			adj[a - 1].add(b - 1);
			adj[b - 1].add(a - 1);
		}
		ans = new int[farms];
		ans[0] = 1;
		dfs(0, 0);
		int max = 0;
		for (int i : ans) {
			max = Math.max(max, i);
		}
		pw.println(max);
		pw.close();
	}

	public static void dfs(int self, int parent) {
		int len = adj.get(self).size();
		int g = 1;
		for (int i : adj.get(self)) {
			if (i != parent) {
				while (g == ans[self] || g == ans[parent]) g++;
				ans[i] = g;
				dfs(i, self);
				g++;
			}
		}
	}
}
```

> import java.io.*;
> import java.util.*;
> 
> public class Main {
> 	public static int[] ans;
> 	public static List<Integer>[] adj;
> 	public static void main(String[] args) throws IOException {
> 		BufferedReader f = new BufferedReader(new FileReader("planting.in"));
> 		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("planting.out")));
> 		StringTokenizer st = new StringTokenizer(f.readLine());
> 		int farms = Integer.parseInt(st.nextToken());
> 		adj = new List[farms];
> 		for (int i = 0; i < farms; i++) {
> 			adj[i] = new ArrayList<>();
> 		}
> 		for (int i = 0; i < farms - 1; i++) {
> 			st = new StringTokenizer(f.readLine());
> 			int a = Integer.parseInt(st.nextToken());
> 			int b = Integer.parseInt(st.nextToken());
> 			adj[a - 1].add(b - 1);
> 			adj[b - 1].add(a - 1);
> 		}
> 		ans = new int[farms];
> 		ans[0] = 1;
> 		dfs(0, 0);
> 		int max = 0;
> 		for (int i : ans) {
> 			max = Math.max(max, i);
> 		}
> 		pw.println(max);
> 		pw.close();
> 	}
> 
> 	public static void dfs(int self, int parent) {
> 		int len = adj.get(self).size();
> 		int g = 1;
> 		for (int i : adj.get(self)) {
> 			if (i != parent) {
> 				while (g == ans[self] || g == ans[parent]) g++;
> 				ans[i] = g;
> 				dfs(i, self);
> 				g++;
> 			}
> 		}
> 	}
> }

Reviewer 2: ```suggestion
import java.io.*;
import java.util.*;
public class Main {
	public static int[] ans;
	public static List<Integer>[] adj;
	public static void main(String[] args) throws IOException {
		BufferedReader f = new BufferedReader(new FileReader("planting.in"));
		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("planting.out")));
		StringTokenizer st = new StringTokenizer(f.readLine());
		int farms = Integer.parseInt(st.nextToken());
		adj = new List[farms];
		for (int i = 0; i < farms; i++) {
			adj[i] = new ArrayList<>();
		}
		for (int i = 0; i < farms - 1; i++) {
			st = new StringTokenizer(f.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			adj[a - 1].add(b - 1);
			adj[b - 1].add(a - 1);
		}
		ans = new int[farms];
		ans[0] = 1;
		dfs(0, 0);
		int max = 0;
		for (int i : ans) {
			max = Math.max(max, i);
		}
		pw.println(max);
		pw.close();
	}
	public static void dfs(int self, int parent) {
		int len = adj[self].size();
		int g = 1;
		for (int i : adj[self]) {
			if (i != parent) {
				while (g == ans[self] || g == ans[parent]) g++;
				ans[i] = g;
				dfs(i, self);
				g++;
			}
		}
	}
}
```

> 	public static int[] visited;
> 	public static List<Integer>[] adj;
> 	public static void main(String[] args) throws IOException {
> 		BufferedReader f = new BufferedReader(new FileReader("planting.in"));
> 		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("planting.out")));
> 		StringTokenizer st = new StringTokenizer(f.readLine());
> 		int farms = Integer.parseInt(st.nextToken());
> 		adj = new List[farms];
> 		for (int i = 0; i < farms; i++) {
> 			adj[i] = new ArrayList<>();
> 		}
> 		for (int i = 0; i < farms - 1; i++) {
> 			st = new StringTokenizer(f.readLine());
> 			int a = Integer.parseInt(st.nextToken());
> 			int b = Integer.parseInt(st.nextToken());
> 			adj[a - 1].add(b - 1);
> 			adj[b - 1].add(a - 1);
> 		}
> 		visited = new int[farms];
> 		visited[0] = 1;
> 		dfs(0, 0);
> 		int max = 0;
> 		for (int i : visited) {
> 			max = Math.max(max, i);
> 		}
> 		pw.println(max);
> 		pw.close();
> 	}
> 
> 	public static void dfs(int self, int parent) {
> 		int len = adj.get(self).size();
> 		int g = 1;
> 		for (int i : adj.get(self)) {
> 			if (i != parent) {
> 				while (g == visited[self] || g == visited[parent]) g++;
> 				visited[i] = g;
> 				dfs(i, self);
> 				g++;
> 			}
> 		}
> 	}

Reviewer 4: ```suggestion
	public static int[] deg;
	public static List<Integer>[] adj;
	public static void main(String[] args) throws IOException {
		BufferedReader f = new BufferedReader(new FileReader("planting.in"));
		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("planting.out")));
		StringTokenizer st = new StringTokenizer(f.readLine());
		int farms = Integer.parseInt(st.nextToken());
		adj = new List[farms];
		for (int i = 0; i < farms; i++) {
			adj[i] = new ArrayList<>();
		}
		for (int i = 0; i < farms - 1; i++) {
			st = new StringTokenizer(f.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			adj[a - 1].add(b - 1);
			adj[b - 1].add(a - 1);
		}
		deg = new int[farms];
		deg[0] = 1;
		dfs(0, 0);
		int max = 0;
		for (int i : deg) {
			max = Math.max(max, i);
		}
		pw.println(max);
		pw.close();
	}

	public static void dfs(int self, int parent) {
		int len = adj.get(self).size();
		int g = 1;
		for (int i : adj.get(self)) {
			if (i != parent) {
				while (g == deg[self] || g == deg[parent]) g++;
				deg[i] = g;
				dfs(i, self);
				g++;
			}
		}
	}
```

</JavaSection>

</LanguageSection>

## Solution 2 - DFS

**Time Complexity**: $\mathcal{O}(N)$

This can also be solved using DFS. The solution is less elegant, but generates a valid assignment of grass types in the $\texttt{visited}$ array. 

<LanguageSection>

<JavaSection>

```java
import java.io.*;
import java.util.*;

public class Main {
	public static int[] visited;

> 	public static ArrayList<ArrayList<Integer>> adj;

Reviewer 4: ```suggestion
	public static List<Integer>[] adj;
```

	public static void main(String[] args) throws IOException {
		BufferedReader f = new BufferedReader(new FileReader("planting.in"));
		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("planting.out")));
		StringTokenizer st = new StringTokenizer(f.readLine());
		int farms = Integer.parseInt(st.nextToken());

> 		adj = new ArrayList<>();

Reviewer 4: ```suggestion
		adj = new List[farms];
```

		for (int i = 0; i < farms; i++) {

> 			adj.add(new ArrayList<Integer>());

Reviewer 4: ```suggestion
			adj[i] = new ArrayList<>();
```

		}
		for (int i = 0; i < farms - 1; i++) {
			st = new StringTokenizer(f.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());

> 			adj.get(a - 1).add(b - 1);
> 			adj.get(b - 1).add(a - 1);

Reviewer 4: ```suggestion
			adj[a - 1].add(b - 1);
			adj[b - 1].add(a - 1);
```

		}
		visited = new int[farms];

> 		Arrays.fill(visited, 0);

Reviewer 4: ```suggestion
```

</JavaSection>

</LanguageSection>

## Solution 2 - DFS

**Time Complexity**: $\mathcal{O}(N)$

This can also be solved using DFS. The solution is less elegant, but generates a valid assignment of grass types in the $\texttt{visited}$ array. 

<LanguageSection>

<JavaSection>

```java
import java.io.*;
import java.util.*;

public class Main {

> 	public static int[] visited;

Reviewer 4: can we rename this variable?

Author: I guess

	public static ArrayList<ArrayList<Integer>> adj;

> 

Reviewer 4: ```suggestion
```
